File: AlexNet/alexnet.py
# https://github.com/pytorch/vision/blob/master/torchvision/models/alexnet.py
# https://pytorch.org/hub/pytorch_vision_alexnet/

import urllib
import torch
import torch.nn as nn
from PIL import Image
from torchvision import transforms


# AlexNet model
class AlexNet(nn.Module):
    ''' AlexNet architecture implemented in PyTorch '''
    def __init__(self, num_classes : int = 1000) -> None:
        super(AlexNet, self).__init__()

        self.features = nn.Sequential(
            # First Convolutional Layer + Maxpooling
            nn.Conv2d(  # Type of layer that is used to extract features from an image
                in_channels=3,   # Number of channels in the input image. For example in this case it's 3 because the image has colors (RGB -> 3 channels)
                out_channels=64,    # Number of channels produced by the convolutional layer. Can be interpreted as the number of filters used to map the features of the image
                kernel_size=(11, 11),  # Size of the matrix that defines the way pixels are going to be summed together
                stride=4,  # Size of the movement the kernel matrix will do over the image
                padding=2  # Padding of 2 added to the input to make it even
            ),
            nn.ReLU(inplace=True), # Applies the rectified linear unit function (ReLU) in place
            nn.MaxPool2d(  # Type of layer that is used to reduce the size of image tensor
                kernel_size=(3, 3),  # Size of the matrix that defines the region from which pixels are going to be picked
                stride=2  # Size of the movement the kernel matrix will do over the image
            ),
            # Second Convolutional Layer + Maxpooling
            nn.Conv2d(in_channels=64, out_channels=192, kernel_size=(5, 5), stride=1, padding=2),
            nn.ReLU(inplace=True),
            nn.MaxPool2d(kernel_size=(3, 3), stride=2),
            # Third Convolutional Layer
            nn.Conv2d(in_channels=192, out_channels=384, kernel_size=(3, 3), stride=1, padding=1),
            nn.ReLU(inplace=True),
            # Fourth Convolutional Layer
            nn.Conv2d(in_channels=384, out_channels=256, kernel_size=(3, 3), stride=1, padding=1),
            nn.ReLU(inplace=True),
            # Fifth Convolutional Layer + Maxpooling
            nn.Conv2d(in_channels=256, out_channels=256, kernel_size=(3, 3), stride=1, padding=1),
            nn.ReLU(inplace=True),
            nn.MaxPool2d(kernel_size=(3, 3), stride=2),
        )

        # Average Pooling downsamples the data by taking the average over a patch of values, it's adaptive because you define the output size and it will automatically choose kernel and stride values
        self.avgpool = nn.AdaptiveAvgPool2d(output_size=(6, 6))

        self.classifier = nn.Sequential(
            nn.Dropout(0.5),
            nn.Linear(
                in_features=256 * 6 * 6,  # Number of input features. In this case:  256 -> output of last Conv2d, 6*6 -> output size of Adaptive Average Pooling
                out_features=4096  # Number of output features
            ),
            nn.ReLU(),
            nn.Dropout(0.5),
            nn.Linear(in_features=4096, out_features=4096),
            nn.ReLU(inplace=True),
            nn.Linear(in_features=4096, out_features=num_classes)
        )
    
    def forward(self, x : torch.Tensor) -> torch.Tensor:
        x = self.features(x)
        x = self.avgpool(x)
        x = torch.flatten(input=x, start_dim=1) # Flattens the first dimension of the input tensor and therefore returns a 1D Tensor
        x = self.classifier(x)

        return x



# Load pretrained AlexNet model from PyTorch
model = torch.hub.load('pytorch/vision:v0.6.0', 'alexnet', pretrained=True)
# Put model in evaluation mode
model.eval()


# Dowload an example from PyTorch website
url, filename = ("https://github.com/pytorch/hub/raw/master/images/dog.jpg", "dog.jpg")
try:
    urllib.request.URLopener().retrieve(url, filename)
except:
    urllib.request.urlretrieve(url, filename)

# File into PIL image
input_image = Image.open(filename)
# Compose several transformations needed for preprocessing
preprocess = transforms.Compose([
    transforms.Resize(size=256),    #  Resizes the given image to the size passed
    transforms.CenterCrop(size=224), # Crops the given image at the center with a desired output size
    transforms.ToTensor(),  # Converts PIL image or numpy.ndarray to tensor
    transforms.Normalize(   # Normalizes a tensor image with mean and standard deviation
        mean=[0.485, 0.456, 0.406],  # sequence of mean for each channel
        std=[0.229, 0.224, 0.225]  # sequence of standard deviations for each channel
    )
])
# Apply the preprocessing to the image
input_tensor = preprocess(input_image)
# Create a mini-batch as expected by the model
input_batch = input_tensor.unsqueeze(0)

# Move input and model in GPU if available
if torch.cuda.is_available():
    input_batch = input_batch.to('cuda')
    model.to('cuda')

# Prediction
with torch.no_grad():  # Disables the calculation of gradients
    output = model(input_batch)  # Get prediction from model

print(output[0])

# To get probabilities run softmax function on output
probabilities = nn.functional.softmax(output[0], dim=0)
print(probabilities)

File: AlexNet/alexnet.cpp
#include <iostream>
#include <torch/torch.h>

// ----------------------------------- MODEL ----------------------------------- //

struct AlexNetImplementation : torch::nn::Module
{
    AlexNetImplementation(int num_classes=1000)
    : 
        // Convolutions
        conv1(torch::nn::Conv2d(torch::nn::Conv2dOptions(3, 64, (11, 11)).stride(4).padding(2))),
        conv2(torch::nn::Conv2d(torch::nn::Conv2dOptions(64, 192, (5, 5)).stride(1).padding(2))),
        conv3(torch::nn::Conv2d(torch::nn::Conv2dOptions(192, 384, (3, 3)).stride(1).padding(1))),
        conv4(torch::nn::Conv2d(torch::nn::Conv2dOptions(384, 256, (3, 3)).stride(1).padding(1))),
        conv5(torch::nn::Conv2d(torch::nn::Conv2dOptions(256, 256, (3, 3)).stride(1).padding(1))),
        // Maxpooling
        max_pool(torch::nn::MaxPool2d((3, 3), 2)),
        // Adaptive average pooling
        avgpool(torch::nn::AdaptiveAvgPool2d((6, 6))),
        // Linear
        linear(torch::nn::Linear(4096, 4096)),
        linear_out(torch::nn::Linear(4096, num_classes))
    {
        // Reegister all layers
        register_module("conv1", conv1);
        register_module("conv2", conv2);
        register_module("conv3", conv3);
        register_module("conv4", conv4);
        register_module("conv5", conv5);
        register_module("max_pool", max_pool);
        register_module("avgpool", avgpool);
        register_module("linear", linear);
        register_module("linear_out", linear_out);
    }

    // Define forward function to pass the input through the network
    torch::Tensor forward(torch::Tensor x) 
    {
        // First Convolutional Layer + Maxpooling
        x = conv1(x);
        x = torch::relu(x);
        x = max_pool(x);
        // Second Convolutional Layer + Maxpooling
        x = conv2(x);
        x = torch::relu(x);
        x = max_pool(x);
        // Third Convolutional Layer + Maxpooling
        x = conv3(x);
        x = torch::relu(x);
        // Fourth Convolutional Layer + Maxpooling
        x = conv4(x);
        x = torch::relu(x);
        // Fifth Convolutional Layer + Maxpooling
        x = conv5(x);
        x = torch::relu(x);
        x = max_pool(x);
        // Adaptive average poolin
        x = avgpool(x);
        // Flatten the input
        x = torch::flatten(x, 1);
        // Fully connected network with dropout
        x = torch::dropout(x, 0.5, is_training());
        x = linear(x);
        x = torch::relu(x);
        x = torch::dropout(x, 0.5, is_training());
        x = linear(x);
        x = torch::relu(x);
        x = linear_out(x);

        return x;
    }

    // Module layers
    torch::nn::Conv2d conv1, conv2, conv3, conv4, conv5; 
    torch::nn::MaxPool2d max_pool;
    torch::nn::AdaptiveAvgPool2d avgpool;
    torch::nn::Linear linear, linear_out;
};
TORCH_MODULE_IMPL(AlexNet, AlexNetImplementation);   // Wraps AlexNetImplementation to AlexNet with a shared_ptr and abstracts away any memory management

int main() {
    // Device -> set CUDA if available
    torch::Device device = torch::kCPU;
    if (torch::cuda::is_available()) {
        torch::Device device = torch::kCUDA;
        std::cout << "CUDA available" << std::endl;
    }

    // Init model
    auto alexnet = AlexNet();
    // Model on CUDA (if available, CPU otherwise)
    alexnet->to(device);

    // Optimizer
    torch::optim::Adam optimizer(alexnet->parameters(), torch::optim::AdamOptions(1e-3));

    // Dummy data
    int batch_size = 128;
    torch::Tensor x_train = torch::ones({batch_size, 3, 224, 224}, device);   // Input data
    torch::Tensor y_train = torch::randn({batch_size, 1000}, device);    // Target data



    // ----------------------------------- TRAINING ----------------------------------- //

    int EPOCHS = 1000;

    // Model in training mode
    alexnet->train();

    for (int i = 0; i < EPOCHS; i++) {
        // Gradients to zero
        optimizer.zero_grad();
        // Get predictions from model
        torch::Tensor output = alexnet->forward(x_train);
        // Calculate loss
        torch::Tensor loss = torch::mse_loss(output, y_train);
        // Calculate gradients
        loss.backward();
        // Update parameters
        optimizer.step();

        if (i % 10 == 0)
            std::cout << "Loss: " << loss << std::endl;
    }



    // ----------------------------------- TESTING ----------------------------------- //

    // Dummy data
    torch::Tensor x_test = torch::randn({batch_size, 3, 224, 224}, device);
    torch::Tensor y_test = torch::randn({batch_size, 1000}, device);

    // Model in validation mode
    alexnet->eval();

    // Get predictions from model
    torch::Tensor out_test = alexnet->forward(x_test);
    // Calculate the error
    torch::Tensor result = torch::mse_loss(out_test, y_test);

    std::cout << "Test result -> " << result << std::endl;
    // Log the predicted class
    std::cout << "Class -> " << torch::nn::functional::softmax(out_test, torch::nn::functional::SoftmaxFuncOptions(0));
}
